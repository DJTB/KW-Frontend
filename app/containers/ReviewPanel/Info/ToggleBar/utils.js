/**
 * Cycles review info detail level
 * @param  {String} [detailLevel=''] level of detail to show
 * @return {String} new detail level
 */
export function cycleDetailLevel(detailLevel = '') {
  if (detailLevel === 'LOW') return 'MEDIUM';
  if (detailLevel === 'MEDIUM') return 'HIGH';
  return 'LOW';
}
