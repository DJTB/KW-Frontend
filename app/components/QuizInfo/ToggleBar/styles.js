import styled from 'styled-components';
import { darken, lighten } from 'polished';

import { resetButton } from 'shared/styles/utils';
import { transparent, whiteLight, greyLight, greyDark } from 'shared/styles/colors';
import { micro } from 'shared/styles/typography';
import { gutter } from 'shared/styles/layout';
import { shadowBox } from 'shared/styles/shadows';

export const Wrapper = styled.div`
  ${gutter({ position: 'top', mod: 0.25 })}
  ${gutter({ position: 'horizontal', mod: 0.5 })}
  ${gutter({ position: 'bottom' })}
  display: flex;
  z-index: 2;
  list-style: none;
  justify-content: center;
  background-color: ${transparent};
  width: 100%;
`;

export const ToggleButton = styled.button`
  ${resetButton}
  ${micro}
  ${shadowBox}
  ${gutter({ mod: 0.5 })}
  ${gutter({ prop: 'margin', position: 'horizontal', mod: 0.5 })}
  position: relative;
  text-align: center;
  text-transform: uppercase;
  flex: 0 1 300px;
  background-color: ${whiteLight};
  color: ${greyLight};

  ${({ isActive }) => isActive && `
    color: ${greyDark};

    /* Triangle pointer */
    &:after {
      content: '';
      position: absolute;
      bottom: -.6rem;
      left: 50%;
      margin-left: -.9rem;
      border-style: solid;
      border-width: 0 .8rem .8rem .8rem;
      border-color: ${transparent} ${transparent} ${whiteLight} ${transparent};
      z-index: 10;
    }
  `}

  &:hover,
  &:focus {
    color: ${darken(0.1, greyDark)};
    background-color: ${lighten(1, whiteLight)};
    outline: none;
  }

  &:disabled {
    opacity: .7;
    pointer-events: none;
  }
`;
