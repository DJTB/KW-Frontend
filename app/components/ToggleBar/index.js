import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import styled from 'styled-components';
import * as COLORS from 'shared/styles/colors';
import {
  toggleNewSynonymPanel,
  toggleInfoPanels,
  toggleInfoDepth,
 } from 'containers/ReviewInfo/actions';

const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  background-color: rgba(${COLORS.greyLight}, .5);
  color: rgb(${COLORS.blackLight});
`;

const Toggle = styled.div`
  text-align: center;
  padding: .2rem .6rem;
  cursor: pointer;
  background-color: rgb(${COLORS.orange});
`;

function ToggleBar({ _toggleNewSynonymPanel, _toggleInfoPanels, _toggleInfoDepth }) {
  return (
    <Wrapper>
      <Toggle onClick={_toggleInfoDepth}>Toggle Detail</Toggle>
      <Toggle onClick={_toggleInfoPanels}>Toggle Info Panel</Toggle>
      <Toggle onClick={_toggleNewSynonymPanel}>Add New Synonym</Toggle>
    </Wrapper>
  );
}

ToggleBar.propTypes = {
  _toggleNewSynonymPanel: PropTypes.func.isRequired,
  _toggleInfoPanels: PropTypes.func.isRequired,
  _toggleInfoDepth: PropTypes.func.isRequired,
};

const mapStateToProps = () => createStructuredSelector({

});

const mapDispatchToProps = (dispatch) => ({
  _toggleNewSynonymPanel: () => dispatch(toggleNewSynonymPanel()),
  _toggleInfoPanels: () => dispatch(toggleInfoPanels()),
  _toggleInfoDepth: () => dispatch(toggleInfoDepth()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ToggleBar);
