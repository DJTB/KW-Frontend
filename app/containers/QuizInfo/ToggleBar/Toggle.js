import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { selectInfoDisabled } from 'containers/QuizPage/selectors';
import { createStructuredSelector } from 'reselect';

import { ToggleButton } from './styles';

Toggle.propTypes = {
  isActive: PropTypes.bool.isRequired,
  isDisabled: PropTypes.bool.isRequired,
  handleClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
};

function Toggle({ isActive, isDisabled, handleClick, children }) {
  return (
    <ToggleButton
      tabIndex="0"
      onClick={handleClick}
      isActive={isActive}
      disabled={isDisabled} // html attr -> css:pseudo
    >
      {children}
    </ToggleButton>
  );
}

const mapStateToProps = createStructuredSelector({
  isDisabled: selectInfoDisabled,
});

export default connect(mapStateToProps)(Toggle);
