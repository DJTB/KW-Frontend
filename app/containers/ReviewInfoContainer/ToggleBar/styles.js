import styled from 'styled-components';
import { media } from 'shared/styles/media';
import { siteMaxWidth } from 'shared/styles/sizing';
import { resetButton } from 'shared/styles/utils';
import { transparent, whiteLight, greyLight, greyDark } from 'shared/styles/colors';
import { micro } from 'shared/styles/typography';
import { shadowBox } from 'shared/styles/shadows';

export const Wrapper = styled.div`
  display: flex;
  z-index: 2;
  margin: 0;
  list-style: none;
  justify-content: center;
  background-color: ${transparent};
  padding: 0 .4rem .4rem;
  width: 100%;
  ${media().sm`
    padding-left: .2rem; /* toggles have .2rem side margins */
    padding-right: .2rem;
    max-width: ${siteMaxWidth}px;
    margin-left: auto;
    margin-right: auto;
  `}
`;

export const ToggleButton = styled.button`
  ${resetButton}
  ${micro}
  ${shadowBox}
  position: relative;
  text-align: center;
  text-transform: uppercase;
  flex: 0 1 300px;
  ${''/* cursor: pointer; */}
  padding: .2rem;
  margin: 0 .2rem;
  background-color: ${whiteLight};
  color: ${greyLight};

  ${({ isActive }) => isActive && `
    color: ${greyDark};

    /* Triangle pointer */
    &:after {
      content: '';
      position: absolute;
      bottom: -.6rem;
      left: 50%;
      margin-left: -.9rem;
      border-style: solid;
      border-width: 0 .8rem .8rem .8rem;
      border-color: ${transparent} ${transparent} ${whiteLight} ${transparent};
      z-index: 10;
    }
  `}

  &:hover {
    color: ${greyDark};
  }

  &:disabled {
    opacity: .7;
    cursor: not-allowed;
    pointer-events: none;
  }
`;
