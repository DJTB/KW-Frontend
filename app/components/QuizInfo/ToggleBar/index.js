import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import quiz from 'pages/QuizPage/actions';
import {
  selectInfoActivePanel,
  selectInfoDetailLevel,
} from 'pages/QuizPage/selectors';
import { Wrapper } from './styles';
import Toggle from './Toggle';

class ToggleBar extends React.Component {
  static propTypes = {
    activePanel: PropTypes.string,
    updateInfo: PropTypes.func.isRequired,
    cycleInfoDetail: PropTypes.func.isRequired,
    detailLevel: PropTypes.number.isRequired,
    isDisabled: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    activePanel: 'INFO',
  };

  onNotesClick = () => this.props.updateInfo({ activePanel: 'NOTES' });
  onInfoClick = () => {
    if (this.props.activePanel === 'INFO') {
      this.props.cycleInfoDetail();
    } else {
      this.props.updateInfo({ activePanel: 'INFO' });
    }
  };
  onSynonymClick = () => this.props.updateInfo({ activePanel: 'SYNONYM' });

  infoPanelText = (level, isDisabled, isActive) => {
    const text = 'Info';
    if (isActive && !isDisabled) {
      if (level === 0) return `${text}: LOW`;
      if (level === 1) return `${text}: MID`;
      return `${text}: HIGH`;
    }
    return text;
  };

  render() {
    const { activePanel, detailLevel, isDisabled } = this.props;
    return (
      <Wrapper>
        <Toggle
          isActive={activePanel === 'NOTES'}
          handleClick={this.onNotesClick}
        >
          Notes
        </Toggle>
        <Toggle
          innerRef={(node) => {
            this.infoRef = node;
          }}
          isActive={activePanel === 'INFO'}
          handleClick={this.onInfoClick}
        >
          {this.infoPanelText(detailLevel, isDisabled, activePanel === 'INFO')}
        </Toggle>
        <Toggle
          isActive={activePanel === 'SYNONYM'}
          handleClick={this.onSynonymClick}
        >
          Add Synonym
        </Toggle>
      </Wrapper>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  activePanel: selectInfoActivePanel,
  detailLevel: selectInfoDetailLevel,
});

const mapDispatchToProps = {
  updateInfo: quiz.info.update,
  cycleInfoDetail: quiz.info.cycledetail,
};

export default connect(mapStateToProps, mapDispatchToProps)(ToggleBar);
