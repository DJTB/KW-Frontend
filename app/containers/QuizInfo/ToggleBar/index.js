import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import quiz from 'containers/QuizPage/actions';
import { selectInfoActivePanel, selectInfoDetailLevel } from 'containers/QuizPage/selectors';
import { Wrapper } from './styles';
import Toggle from './Toggle';


class ToggleBar extends React.Component {
  static propTypes = {
    activePanel: PropTypes.string,
    updateInfo: PropTypes.func.isRequired,
    cycleInfoDetail: PropTypes.func.isRequired,
    isDisabled: PropTypes.bool.isRequired,
    detailLevel: PropTypes.number.isRequired,
  };

  static defaultProps = {
    activePanel: 'INFO',
  };

  componentDidUpdate(prevProps) {
    if (prevProps.isDisabled && !this.props.isDisabled) {
      this.infoRef.focus();
    }
  }

  onNotesClick = () => this.props.updateInfo({ activePanel: 'NOTES' });
  onInfoClick = () => {
    if (this.props.activePanel === 'INFO') {
      this.props.cycleInfoDetail();
    } else {
      this.props.updateInfo({ activePanel: 'INFO' });
    }
  };
  onSynonymClick = () => this.props.updateInfo({ activePanel: 'SYNONYM' });

  render() {
    const { activePanel, detailLevel } = this.props;
    return (
      <Wrapper>
        <Toggle isActive={activePanel === 'NOTES'} handleClick={this.onNotesClick}>Notes</Toggle>
        <Toggle
          innerRef={(node) => { this.infoRef = node; }}
          isActive={activePanel === 'INFO'}
          handleClick={this.onInfoClick}
        >
          {`Info: ${detailLevel ? 'HIGH' : 'LOW'}`}
        </Toggle>
        <Toggle isActive={activePanel === 'SYNONYM'} handleClick={this.onSynonymClick}>Add Synonym</Toggle>
      </Wrapper>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  activePanel: selectInfoActivePanel,
  detailLevel: selectInfoDetailLevel,
});

const mapDispatchToProps = {
  updateInfo: quiz.info.update,
  cycleInfoDetail: quiz.info.cycledetail,
};

export default connect(mapStateToProps, mapDispatchToProps)(ToggleBar);
