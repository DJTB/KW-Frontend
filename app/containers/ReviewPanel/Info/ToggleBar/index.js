import React from 'react';
import PropTypes from 'prop-types';

import titleCase from 'voca/title_case';
import { cycleDetailLevel } from './utils';
import { Wrapper } from './styles';
import Toggle from './Toggle';

const DETAIL_LEVELS = {
  LOW: 'LOW',
  MEDIUM: 'MEDIUM',
  HIGH: 'HIGH',
};

class ToggleBar extends React.Component {
  static propTypes = {
    detailLevel: PropTypes.oneOf(Object.values(DETAIL_LEVELS)),
    isDisabled: PropTypes.bool,
    notes: PropTypes.shape({ isActive: PropTypes.bool }),
    info: PropTypes.shape({ isActive: PropTypes.bool }),
    synonym: PropTypes.shape({ isActive: PropTypes.bool }),
    showNotesPanel: PropTypes.func.isRequired,
    showSynonymPanel: PropTypes.func.isRequired,
    showInfoPanel: PropTypes.func.isRequired,
  }

  static defaultProps = {
    detailLevel: DETAIL_LEVELS.LOW,
    isDisabled: false,
    notes: { isActive: false },
    info: { isActive: true },
    synonym: { isActive: false },
  };

  state = {
    // start with user settings, then manage with state when mounted
    detailLevel: this.props.detailLevel,
  }

  handleNotesClick = () => {
    this.props.showNotesPanel();
  }

  handleInfoClick = () => {
    if (this.props.info.isActive) {
      this.setState({ detailLevel: cycleDetailLevel(this.state.detailLevel) });
    } else {
      this.props.showInfoPanel();
    }
  }

  handleSynonymClick = () => {
    this.props.showSynonymPanel();
  }

  render() {
    const { isDisabled, notes, info, synonym } = this.props;
    return (
      <Wrapper>
        <Toggle
          isActive={notes.isActive}
          isDisabled={isDisabled}
          handleClick={this.handleNotesClick}
        >
          Notes
        </Toggle>
        <Toggle
          isActive={info.isActive}
          isDisabled={isDisabled}
          handleClick={this.handleInfoClick}
        >
          Info: {titleCase(this.state.detailLevel)}
        </Toggle>
        <Toggle
          isActive={synonym.isActive}
          isDisabled={isDisabled}
          handleClick={this.handleSynonymClick}
        >
          New Synonym
        </Toggle>
      </Wrapper>
    );
  }
}

/* const mapStateToProps = createStructuredSelector({
  isAnswerMarked: selectAnswerMarked,
  panelToShow: selectPanelToShow,
  detailLevel: selectDetailLevel,
});

const mapDispatchToProps = (dispatch) => ({
  showNotesPanel: () => dispatch(actions.showPanel(REVIEW_INFO_PANELS.NOTES)),
  showSynonymPanel: () => dispatch(actions.showPanel(REVIEW_INFO_PANELS.SYNONYM)),
  showInfoPanel: () => dispatch(actions.showPanel(REVIEW_INFO_PANELS.INFO)),
  cycleInfoDetail: () => dispatch(globalActions.cycleInfoDetail()),
});
*/
export default ToggleBar;
