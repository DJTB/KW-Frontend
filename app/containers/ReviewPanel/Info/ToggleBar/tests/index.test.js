import React from 'react';
import { shallow } from 'enzyme';
import ToggleBar from '../index';

describe('<ToggleBar />', () => {
  it('should match baseline snapshot', () => {
    const renderedComponent = shallow(
      <ToggleBar
        showNotesPanel={jest.fn()}
        showSynonymPanel={jest.fn()}
        showInfoPanel={jest.fn()}
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should adopt optional props', () => {
    const renderedComponent = shallow(
      <ToggleBar
        showNotesPanel={jest.fn()}
        showSynonymPanel={jest.fn()}
        showInfoPanel={jest.fn()}
        detailLevel="HIGH"
        isDisabled
        notes={{ isActive: true }}
        info={{ isActive: false }}
        synonym={{ isActive: true }}
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });
});
