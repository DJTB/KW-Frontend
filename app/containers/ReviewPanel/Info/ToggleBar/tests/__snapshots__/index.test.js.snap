// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleBar /> should adopt optional props 1`] = `
<styled.section>
  <Toggle
    handleClick={[Function]}
    isActive={true}
    isDisabled={true}
  >
    Notes
  </Toggle>
  <Toggle
    handleClick={[Function]}
    isActive={false}
    isDisabled={true}
  >
    Info: 
    High
  </Toggle>
  <Toggle
    handleClick={[Function]}
    isActive={true}
    isDisabled={true}
  >
    New Synonym
  </Toggle>
</styled.section>
`;

exports[`<ToggleBar /> should match baseline snapshot 1`] = `
<styled.section>
  <Toggle
    handleClick={[Function]}
    isActive={false}
    isDisabled={false}
  >
    Notes
  </Toggle>
  <Toggle
    handleClick={[Function]}
    isActive={true}
    isDisabled={false}
  >
    Info: 
    Low
  </Toggle>
  <Toggle
    handleClick={[Function]}
    isActive={false}
    isDisabled={false}
  >
    New Synonym
  </Toggle>
</styled.section>
`;
